### PR TITLE
Fix SR-5031

### PR DESCRIPTION
**Pull request by norio-nomura**
- Add `Manifest.versionSpecificManifest(path:fileSystem:)`
- Use `Manifest.versionSpecificManifest(path:fileSystem:)` instead of manipulating path with `Manifest.filename`
- Remove `ToolsVersion.toolsVersionFileName`

### CODE REVIEW

File: Sources/Commands/SwiftTool.swift [hunk 1]
```diff
@@ -455,7 +455,7 @@ private func findPackageRoot() -> AbsolutePath? {
     // FIXME: It would be nice to move this to a generalized method which takes path and predicate and
     // finds the lowest path for which the predicate is true.
     var root = currentWorkingDirectory
-    while !isFile(root.appending(component: Manifest.filename)) {
+    while !isFile(Manifest.versionSpecificManifest(path: root)) {
```
Comment (aciidb0mb3r) [inline Sources/Commands/SwiftTool.swift]: We should require that `Package.swift` be present at top level. So please revert this line.

File: Sources/PackageLoading/ManifestLoader.swift [hunk 2]
```diff
@@ -185,7 +172,8 @@ public final class ManifestLoader: ManifestLoaderProtocol {
         guard baseURL.chuzzle() != nil else { fatalError() }  //TODO
 
         // Compute the actual input file path.
-        let path: AbsolutePath = isDirectory(inputPath) ? inputPath.appending(component: Manifest.filename) : inputPath
+        let path: AbsolutePath = isDirectory(inputPath) ?
+            Manifest.versionSpecificManifest(path: inputPath, fileSystem: fileSystem) : inputPath
```
Comment (aciidb0mb3r) [inline Sources/PackageLoading/ManifestLoader.swift]: I think we should just remove this and use the `inputPath` directly now, i.e., it is illegal to provide a directory path to this method.

File: Sources/PackageLoading/ToolsVersionLoader.swift [hunk 3]
```diff
@@ -36,15 +36,15 @@ public class ToolsVersionLoader: ToolsVersionLoaderProtocol {
         public var description: String {
             switch self {
             case .malformed(let versionSpecifier, let file):
-                let file = file.appending(component: ToolsVersion.toolsVersionFileName)
+                let file = Manifest.versionSpecificManifest(path: file)
```
Comment (aciidb0mb3r) [inline Sources/PackageLoading/ToolsVersionLoader.swift]: Instead of re-computing here, we can directly pass the file path from `load(at:fileSystem)` method

File: Sources/PackageLoading/ToolsVersionLoader.swift [hunk 4]
```diff
@@ -36,15 +36,15 @@ public class ToolsVersionLoader: ToolsVersionLoaderProtocol {
...
                 return "The version specifier '\(versionSpecifier)' in '\(file.asString)' is not valid"
             }
         }
     }
 
     public func load(at path: AbsolutePath, fileSystem: FileSystem) throws -> ToolsVersion {
         // The file which contains the tools version.
-        let file = path.appending(component: ToolsVersion.toolsVersionFileName)
+        let file = Manifest.versionSpecificManifest(path: path, fileSystem: fileSystem)
```
Comment (aciidb0mb3r) [inline Sources/PackageLoading/ToolsVersionLoader.swift]: It is a little unfortunate that we will compute this twice...maybe we should lift it out and let clients handle it. But this is fine for now.

File: Sources/PackageModel/Manifest.swift [hunk 5]
```diff
@@ -27,6 +28,18 @@ public final class Manifest: ObjectIdentifierProtocol, CustomStringConvertible {
     /// The standard filename for the manifest.
     public static var filename = basename + ".swift"
 
+    /// Version Specific Manifest in path
+    public static func versionSpecificManifest(path: AbsolutePath, fileSystem: FileSystem? = nil) -> AbsolutePath {
```
Comment (aciidb0mb3r) [inline Sources/PackageModel/Manifest.swift]: Can you rename `path` to `packagePath`.

File: Sources/PackageModel/Manifest.swift [hunk 6]
```diff
@@ -27,6 +28,18 @@ public final class Manifest: ObjectIdentifierProtocol, CustomStringConvertible {
     /// The standard filename for the manifest.
     public static var filename = basename + ".swift"
 
+    /// Version Specific Manifest in path
```
Comment (aciidb0mb3r) [inline Sources/PackageModel/Manifest.swift]: Please briefly describe what the function does.

File: Sources/TestSupport/MockDependencyGraph.swift [hunk 7]
```diff
@@ -141,7 +141,7 @@ public struct MockManifestGraph {
 
         // Create the root manifest.
         rootManifest = Manifest(
-            path: path.appending(component: Manifest.filename),
+            path: Manifest.versionSpecificManifest(path: path, fileSystem: inMemory?.fs),
```
Comment (aciidb0mb3r) [inline Sources/TestSupport/MockDependencyGraph.swift]: I don't think we need to change this.

File: Sources/TestSupport/MockDependencyGraph.swift [hunk 8]
```diff
@@ -153,7 +153,7 @@ public struct MockManifestGraph {
         var manifests = Dictionary(items: packages.map({ package -> (MockManifestLoader.Key, Manifest) in
             let url = repos[package.name]!.url
             let manifest = Manifest(
-                path: AbsolutePath(url).appending(component: Manifest.filename),
+                path: Manifest.versionSpecificManifest(path: AbsolutePath(url), fileSystem: inMemory?.fs),
```
Comment (aciidb0mb3r) [inline Sources/TestSupport/MockDependencyGraph.swift]: Same, this doesn't need to be changed.

File: Sources/TestSupport/misc.swift [hunk 9]
```diff
@@ -235,7 +235,7 @@ public func loadMockPackageGraph4(
     var rootManifest: Manifest!
     for (url, package) in packageMap {
         let manifest = Manifest(
-            path: AbsolutePath(url).appending(component: Manifest.filename),
+            path: Manifest.versionSpecificManifest(path: AbsolutePath(url), fileSystem: fs),
```
Comment (aciidb0mb3r) [inline Sources/TestSupport/misc.swift]: Same, these shouldn't need to be changed.

File: Sources/Workspace/ToolsVersionWriter.swift [hunk 10]
```diff
@@ -19,7 +19,7 @@ import Utility
 ///   - path: The path of the package.
 ///   - version: The version to write.
 public func writeToolsVersion(at path: AbsolutePath, version: ToolsVersion, fs: inout FileSystem) throws {
-    let file = path.appending(component: ToolsVersion.toolsVersionFileName)
+    let file = Manifest.versionSpecificManifest(path: path, fileSystem: fs)
```
Comment (aciidb0mb3r) [inline Sources/Workspace/ToolsVersionWriter.swift]: I think we should always update the main manifest, so please revert this change.

File: Tests/PackageGraphPerformanceTests/PackageGraphPerfTests.swift [hunk 11]
```diff
@@ -38,7 +38,7 @@ class PackageGraphPerfTests: XCTestCasePerf {
             }
             // Create manifest.
             let manifest = Manifest(
-                path: AbsolutePath(url).appending(component: Manifest.filename),
+                path: Manifest.versionSpecificManifest(path: AbsolutePath(url), fileSystem: fs),
```
Comment (aciidb0mb3r) [inline Tests/PackageGraphPerformanceTests/PackageGraphPerfTests.swift]: This isn't needed, please revert.

File: Tests/PackageLoadingTests/ManifestTests.swift [hunk 12]
```diff
@@ -35,7 +35,7 @@ class ManifestTests: XCTestCase {
 
     private func loadManifest(_ contents: ByteString, baseURL: String? = nil) throws -> Manifest {
         let fs = InMemoryFileSystem()
-        let manifestPath = AbsolutePath.root.appending(component: Manifest.filename)
+        let manifestPath = Manifest.versionSpecificManifest(path: AbsolutePath.root, fileSystem: fs)
```
Comment (aciidb0mb3r) [inline Tests/PackageLoadingTests/ManifestTests.swift]: This isn't needed, please revert.

File: Tests/PackageLoadingTests/PD4LoadingTests.swift [hunk 13]
```diff
@@ -29,7 +29,7 @@ class PackageDescription4LoadingTests: XCTestCase {
     ) {
         do {
             let fs = InMemoryFileSystem()
-            let manifestPath = AbsolutePath.root.appending(component: Manifest.filename)
+            let manifestPath = Manifest.versionSpecificManifest(path: AbsolutePath.root, fileSystem: fs)
```
Comment (aciidb0mb3r) [inline Tests/PackageLoadingTests/PD4LoadingTests.swift]: This isn't needed, please revert.

File: Tests/PackageLoadingTests/PackageBuilderTests.swift [hunk 14]
```diff
@@ -1092,7 +1092,8 @@ final class PackageBuilderTester {
     ) {
         let diagnostics = DiagnosticsEngine()
         do {
-            let manifest = Manifest(path: path.appending(component: Manifest.filename), url: "", package: package, version: nil)
+            let manifest = Manifest(path: Manifest.versionSpecificManifest(path: path, fileSystem: fs),
```
Comment (aciidb0mb3r) [inline Tests/PackageLoadingTests/PackageBuilderTests.swift]: This isn't needed, please revert.

File: Tests/WorkspaceTests/WorkspaceTests.swift [hunk 15]
```diff
@@ -1627,7 +1627,8 @@ final class WorkspaceTests: XCTestCase {
                 try localFileSystem.createDirectory(root)
             }
 
-            try localFileSystem.writeFileContents(roots[2].appending(components: Manifest.filename)) { stream in
+            let manifestPath = Manifest.versionSpecificManifest(path: roots[2], fileSystem: localFileSystem)
+            try localFileSystem.writeFileContents(manifestPath) { stream in
                 stream <<< "import PackageDescription" <<< "\n"
                 stream <<< "let package = Package(name: \"root0\")"
```
Comment (aciidb0mb3r) [inline Tests/WorkspaceTests/WorkspaceTests.swift]: These aren;t needed, please revert.

File: Tests/WorkspaceTests/WorkspaceTests.swift [hunk 16]
```diff
@@ -1627,7 +1627,8 @@ final class WorkspaceTests: XCTestCase {
...
             }
```
Comment (aciidb0mb3r) [inline Tests/WorkspaceTests/WorkspaceTests.swift]: We should add a test case that we actually load the correct tools version when version specific manifest is present.

File: Sources/Workspace/ToolsVersionWriter.swift (hunk 10 above)
Comment (norio-nomura) [inline Sources/Workspace/ToolsVersionWriter.swift]: Can I use `Manifest.filename` instead of `ToolsVersion.toolsVersionFileName`? or should I revert removing `ToolsVersion.toolsVersionFileName` too?

Comment (aciidb0mb3r) [inline Sources/Workspace/ToolsVersionWriter.swift]: I think thats fine, we should use `Manifest.filename`

**Comment by norio-nomura** [comment]
Updated!

**Comment by norio-nomura** [comment]
Once this PR is merged, I will also make a PR for swift-4.0-branch.

**Comment by aciidb0mb3r** [comment]
@norio-nomura We don't need PR for swift-4.0-branch as we're going to re-branch from master soon.

**Comment by aciidb0mb3r** [comment]
@swift-ci please smoke test

**Comment by aciidb0mb3r** [comment]
@swift-ci please smoke test

**Comment by norio-nomura** [comment]
> We don't need PR for swift-4.0-branch as we're going to re-branch from master soon.

Oh, I see. 👍 [[[swift-dev] final re-branching for Swift 4 is June 1](https://lists.swift.org/pipermail/swift-dev/Week-of-Mon-20170417/004405.html)

**Comment by aciidb0mb3r** [comment]
@swift-ci please smoke test

**Comment by aciidb0mb3r** [comment]
@swift-ci please smoke test

**Comment by norio-nomura** [comment]
Thanks! :pray:
